### PR TITLE
[WebGPU] Patches for a few WebGPU runtime errors. 

### PR DESCRIPTION
**Pull request by ScottTodd**
Getting closer to being able to run programs like posenet.

* Workaround for https://github.com/iree-org/iree/issues/11321 by setting `--iree-codegen-gpu-native-math-precision=true` in WebGPU samples
* Workaround for errors like
  ```
  [Buffer] usage (BufferUsage::(MapRead|CopyDst)) doesn't include BufferUsage::CopySrc.

  Buffer usages (BufferUsage::(MapRead|CopySrc|CopyDst)) is invalid. If a buffer usage contains BufferUsage::MapRead the only other allowed usage is BufferUsage::CopyDst.
  ```
  simple_allocator will assume CopySrc, application code can directly allocate CopyDst. We should model CopySrc/CopyDst in the compiler based on usage and then support allocating buffers with either usage.
* Fix for errors like
  ```
  No bind group set at index 1.
   - While encoding [ComputePassEncoder].DispatchWorkgroups(3, 65, 89).
  ```

I see various errors when running under ASan, both before and after these changes. I expect I'll be able to better address them when using proper async readback/execution (integrating with the browser event loop).

skip-ci: no WebGPU HAL CI coverage yet

### CODE REVIEW

**Comment by ScottTodd** [comment]
I'm also still carrying forward a few other patches not included here:

* `iree_hal_heap_buffer_wrap(`: skip the `!iree_host_size_has_alignment` check (memory from WebGPU does not meet our alignment requirements)
* Comment out `wgpuBufferDestroy(buffer->handle);` to leave cleanup to WebGPU. Seeing errors like `[Buffer] used in submit while destroyed`

**Review by benvanik — CHANGES_REQUESTED** [review]
🤦 had reviewed this last week but never submitted it ;(

File: experimental/web/sample_webgpu/main.c [hunk 1]
```diff
@@ -282,7 +283,7 @@ static iree_status_t parse_inputs_into_call(
 
 typedef struct iree_buffer_map_userdata_t {
   iree_hal_buffer_view_t* buffer_view;
```
[the remark this answers is not in the document]
Comment (ScottTodd) [reply experimental/web/sample_webgpu/main.c]: Renamed `buffer_view` to `source_buffer_view`. This struct is passed in to `wgpuBufferMapAsync`, so `iree_buffer_map_userdata_t` still makes sense I think.

File: experimental/web/sample_webgpu/main.c [hunk 2]
```diff
@@ -515,6 +548,7 @@ static iree_status_t print_outputs_from_call(
       if (iree_hal_buffer_view_isa(variant.ref)) {
         iree_hal_buffer_view_t* buffer_view =
             iree_hal_buffer_view_deref(variant.ref);
+        iree_hal_buffer_view_retain(buffer_view);  // Release in async callback.
```
[the remark this answers is not in the document]
Comment (ScottTodd) [reply experimental/web/sample_webgpu/main.c]: Ahhh good catch. I had similar thoughts last night. Moved to where the struct stores the pointer.

File: runtime/src/iree/hal/drivers/webgpu/command_buffer.c [hunk 3]
```diff
@@ -850,6 +850,15 @@ static iree_status_t iree_hal_webgpu_command_buffer_prepare_dispatch(
     command_buffer->state.bind_groups[i].handle = handle;
   }
 
+  // Pad up to IREE_HAL_WEBGPU_PARAMS_BIND_GROUP_INDEX with empty bind groups.
+  WGPUBindGroup empty_handle = command_buffer->staging_buffer->empty_bind_group;
```
[the remark this answers is not in the document]
Comment (ScottTodd) [reply runtime/src/iree/hal/drivers/webgpu/command_buffer.c]: Added bitfield and set/unset/check code. Also checked for push_constant_count > 0.

File: experimental/web/sample_webgpu/main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (benvanik) [reply experimental/web/sample_webgpu/main.c]: More about the map API being used to do a specific thing (readback), and if you added another use of wgpuBufferMapAsync for upload (or something else) you'd likely need a different struct for it (source buffer -> target buffer view). But NBD :)

File: runtime/src/iree/hal/buffer_heap.c [hunk 4]
```diff
@@ -185,6 +185,7 @@ iree_status_t iree_hal_heap_buffer_wrap(
   IREE_ASSERT_ARGUMENT(out_buffer);
   IREE_TRACE_ZONE_BEGIN(z0);
 
+#if !defined(IREE_PLATFORM_EMSCRIPTEN)  // WebGPU buffer data isn't aligned?
```
Comment (benvanik) [inline runtime/src/iree/hal/buffer_heap.c:188]: this is a bit too sketchy (and applies to CPU in emscripten, which isn't what we want)

we could add a iree_hal_memory_access_t bit like IREE_HAL_MEMORY_ACCESS_UNALIGNED and disable this check when set (and other checks we may perform in other places)

the calling code wrapping the buffer can then OR in that flag if it knows it's ok to do so

**Review by benvanik — CHANGES_REQUESTED** [review]
one more thing then LGTM!

File: runtime/src/iree/hal/buffer_heap.c (hunk 4 above)
Comment (ScottTodd) [reply runtime/src/iree/hal/buffer_heap.c:188]: Okay, I'll drop that from this PR and keep the patch locally until I implement a better fix (like what you just suggested).

**Comment by ScottTodd** [comment]
> I see various errors when running under ASan, both before and after these changes. I expect I'll be able to better address them when using proper async readback/execution (integrating with the browser event loop).

Debugging this a bit... something about the async callback is tripping up ASan, and I'm not sure if it's a false positive.
```
iree_api_webgpu.js:76 (C) Direct leak of 36 byte(s) in 1 object(s) allocated from:
iree_api_webgpu.js:76 (C)     #0 0x188932 in __sanitizer::StackTrace::GetCurrentPc()+0x188932 (http://localhost:8000/web-sample-webgpu.wasm+0x188932)
iree_api_webgpu.js:76 (C)     #1 0x1592f3 in calloc+0x1592f3 (http://localhost:8000/web-sample-webgpu.wasm+0x1592f3)
iree_api_webgpu.js:76 (C)     #2 0x101d1a in iree_allocator_system_ctl+0x101d1a (http://localhost:8000/web-sample-webgpu.wasm+0x101d1a)
iree_api_webgpu.js:76 (C)     #3 0x100bd9 in iree_allocator_malloc+0x100bd9 (http://localhost:8000/web-sample-webgpu.wasm+0x100bd9)
iree_api_webgpu.js:76 (C)     #4 0xa2643 in iree_vm_list_create+0xa2643 (http://localhost:8000/web-sample-webgpu.wasm+0xa2643)
iree_api_webgpu.js:76 (C)     #5 0x2a318 in iree_runtime_call_initialize+0x2a318 (http://localhost:8000/web-sample-webgpu.wasm+0x2a318)
iree_api_webgpu.js:76 (C)     #6 0x2aab1 in iree_runtime_call_initialize_by_name+0x2aab1 (http://localhost:8000/web-sample-webgpu.wasm+0x2aab1)
```

**Comment by benvanik** [comment]
Interesting - wonder if there's some enter/exit ASAN runtime goo required to let it know about the stack?

**Comment by ScottTodd** [comment]
I scanned through https://emscripten.org/docs/debugging/Sanitizers.html but didn't see any callouts. There's definitely at least one real issue that I'm trying to track down. Execution is correct for programs without push constant emulation, but more complex programs (which use push constant emulation, among other things) use invalid memory or something... maybe I can find a minimal program that uses push constants to help narrow it down.

I see those ASan "leaks" with even simple programs.

**Comment by benvanik** [comment]
could try a simple sample that purposefully mallocs in the callback and then one that mallocs and frees in the callback - can try to isolate if it's an allocation crossing the callback boundary or all allocations related to the callback.

**Comment by ScottTodd** [comment]
Rebased and fixed a few things, PTAL.

The branch does still has issues running programs. sample_abs is broken now with:
`(C) iree/runtime/src/iree/hal/command_buffer_validation.c:81: PERMISSION_DENIED; requested buffer usage is not supported for the buffer on this queue; buffer allows TRANSFER|DISPATCH_STORAGE|MAPPING, operation requires TRANSFER_SOURCE (allocator compatibility mismatch)`

I thought that worked before, so there may have been some usage changes on `main`...

**Comment by benvanik** [comment]
huh, maybe that's a bug in the validation - I'd expect TRANSFER to cover TRANSFER_SOURCE - lemme check

**Comment by benvanik** [comment]
oh, may be the webgpu iree_hal_allocator_query_buffer_compatibility implementation

**Comment by ScottTodd** [comment]
> oh, may be the webgpu iree_hal_allocator_query_buffer_compatibility implementation

Good call. I'll apply https://github.com/openxla/iree/pull/11599 (that works, next error is the `iree/runtime/src/iree/hal/buffer_heap.c:194: OUT_OF_RANGE; imported heap buffer data must be aligned to 64; got 0x3d3e8` that I had a local hack/patch for...)

File: runtime/src/iree/hal/buffer_heap.c (hunk 4 above)
Comment (ScottTodd) [reply runtime/src/iree/hal/buffer_heap.c]: Added `IREE_HAL_MEMORY_ACCESS_UNALIGNED`, simple_abs runs again without local patches.

**Comment by ScottTodd** [comment]
simple_abs: runs correctly, with readback -> output result

fullyconnected:
```
Writable storage buffer binding aliasing found between [BindGroup] set at bind group index 0, binding index 0, and [BindGroup] set at bind group index 0, binding index 2, with overlapping ranges (offset: 0, size: 256) and (offset: 0, size: 256) in [Buffer].
 - While encoding [ComputePassEncoder].DispatchWorkgroups(1, 5, 1).
 
[Invalid CommandBuffer] is invalid.
    at ValidateObject (../../third_party/dawn/src/dawn/native/Device.cpp:689)
    at ValidateSubmit (../../third_party/dawn/src/dawn/native/Queue.cpp:442)
```

mobilebert/posenet/mobilessd: not tested yet, need to download input models and compile on-demand to test

**Comment by benvanik** [comment]
that's catastrophically bad - webgpu sadness ;(

**Review by benvanik — APPROVED** [review]

**Pull request merged**